server/src/main/java/org/candlepin/controller/QpidStatusMonitor.java

```diff
@@ -0,0 +1,152 @@
+        if (delay < 1) {
+            int defaultDelay = Integer.parseInt(
+                ConfigProperties.DEFAULT_PROPERTIES.get(ConfigProperties.QPID_MODE_TRANSITIONER_DELAY));
+            log.warn("{} is an invalid delay setting. Must be greater than 0. Defaulting to {}", delay,
+                defaultDelay);
+            delay = defaultDelay;
+        }
+        this.lastStatus = QpidStatus.DOWN;
+    }
+
+    /**
+     * Other dependencies are injected using method injection so
+     * that Guice can handle circular dependencies
+     */
+    @Inject
+    public void setQmf(QpidQmf qmf) {
+        this.qmf = qmf;
+    }
+
+    @Inject
+    public void setQpidConnection(QpidConnection qpidConnection) {
+        this.qpidConnection = qpidConnection;
+    }
+
+    /**
+     * Executes a single monitoring check.
+     */
+    @Override
+    public void run() {
+        try {
+            log.debug("Executing Qpid status check...");
+            monitor();
+        }
+        finally {
+            log.debug("Next check will be in {} seconds.", delay);
+        }
+    }
+
+    /**
+     * Adds a listener that will be notified when the connection status changes.
+     *
+     * @param listener the listener to be added.
+     */
+    public void addStatusChangeListener(QpidStatusListener listener) {
+        this.listeners.add(listener);
+    }
+
+    /**
+     * Starts/schedules the connection monitoring process. The check will be run at a configured
+     * interval.
+     *
+     * @see ConfigProperties
+     */
+    public void schedule() {
+        log.info("Starting Qpid Status Monitor. Checks will be performed every {} seconds.", delay);
+        executorService.scheduleWithFixedDelay(this, 10, delay, TimeUnit.SECONDS);
+    }
+
+    /**
+     * Monitors the Qpid connection and notifies all listeners when the connection status changes.
+     * If a listener throws an exception, the exception is logged and the next listener is notified
+     * in turn.
+     *
+     * This method should handle all exceptions to ensure that the executor service continues to
+     * schedule the periodic job.
+     */
+    private void monitor() {
+        QpidStatus status;
+        try {
+            status = qmf.getStatus();
+        }
+        catch (Throwable t) {
+            log.error("Error while executing status monitoring check", t);
+            /*
+             * Nothing more we can do here, since this is scheduled thread. We must
+             * hope that this error won't infinitely recur with each scheduled execution
+             */
+            return;
+        }
+
+        log.debug("Qpid connection status - Old: {} New: {}", this.lastStatus, status);
+        try {
+            // If a listener throws an exception, log it, and move to the next listener.
+            // We don't want to skip other listeners just because another failed.
+            fireStatusChange(this.lastStatus, status);
```

Comment:
I'm probably reading this wrong, but the way I read it, this code does the opposite of what the comment says.  In other words, a call is made to `fireStatusChange` which enters a loop.  If any listener in the loop throws an exception, the looping ends, and we proceed to the catch statement.  I would think to match the comment, we would want something like

```java
for (QpidStatusListener listener: this.listeners) {
    try {
        listener.onStatusUpdate(oldStatus, newStatus);
    } catch (Exception e) {
        log.warn("Failed to notify status listener", e);
    }
}
```

Reply:
:+1:  I'll fix that up. Nice catch. Appreciated.